aspnet/fundamentals/routing.rst

```diff
@@ -2,81 +2,201 @@ Routing
+
+This example demonstrates a few more features:
+
+.. code-block:: c#
+
+    routes.MapRoute(
+        name: "blog",
+        template: "/Blog/{*article}",
+        defaults: new { controller = "Blog", action = "ReadArticle" });
+
+This template will match a URL path like ``/Blog/All-About-Routing/Introduction`` and will extract the values ``{ controller = Blog, action = ReadArticle, article = All-About-Routing/Introduction }``. The default route values for ``controller`` and ``action`` are produced by the route even though there are no corresponding route parameters in the template. Default values can be specified in the route template The ``article`` route parameter is defined as a *catch-all* by the appearance of an asterix ``*`` before the route parameter name. Catch-all route parameters capture the remainder of the URL path, and can also match the empty string.
+
+This example adds route constraints and data tokens:
+
+.. code-block:: c#
+
+    routes.MapGet(
+        name: "us_english_products",
+        template: "/en-US/Products/{id}",
+        defaults: new { controller = "Products", action = "Details" },
+        constraints: new { id = new IntConstraint() },
+        dataTokens: new { locale = "en-US" });
+
+This template will match a URL path like ``/en-US/Products/{id}`` and will extract the values ``{ controller = Products, action = Details, id = 5 }`` and the data tokens ``{ locale = en-US }``. 
+
+.. _url-generation:
+
+URL Generation
+^^^^^^^^^^^^^^^
+The ``Route`` class can also perform URL generation by combining a set of route values with its route template. This is logically the reverse process of matching the URL path. 
+
+.. tip:: To better understand URL generation, imagine what URL you want to generate and then think about how a route template would match that URL. What values would be produced? This is the rough equivalent of how URL generation works in the ``Route`` class.
+
+This example uses a basic ASP.NET MVC style route:
+
+.. code-block:: c#
+
+    routes.MapRoute(
+        name: "default",
+        template: "{controller=Home}/{action=Index}/{id?}");
+
+With the route values ``{ controller = Products, action = List }``, this route will generate the URL ``/Products/List``. The route values are substituted for the corresponding route parameters to form the URL path. Since ``id`` is an optional route parameter, it's no problem that it doesn't have a value.
+
+With the route values ``{ controller = Home, action = Index }``, this route will generate the URL ``/``. The route values that were provided match the default values so the segments corresponding to those values can be safely omitted. Note that both would URLs generated would round-trip with this route definition and produce the same route values that were used to generate the URL.
```

Comment:
that I mean by this is that the URL, if matched, would produce a set of route values - if that set of route values were used by this route to generate a URL, it would be the same as the original URL

Feel free to clean this up, but that's what I mean